### ISSUE TITLE
-Warray-bounds warning with asm goto in impossible switch cases

### ISSUE THREAD

**Issue by nathanchance** [llvm/llvm-project#51024]
|  |  |
| --- | --- |
| Bugzilla Link | [51682](https://llvm.org/bz51682) |
| Version | trunk |
| OS | Windows NT |
| Blocks | llvm/llvm-project#4440  |
| Attachments | [False positive -Warray-bounds reproducer](https://user-images.githubusercontent.com/11478138/143762488-bbdccd0e-7783-48dd-afca-9d89c1fdc05f.gz) |
| CC | @DougGregor,@efriedma-quic,@kees,@nickdesaulniers,@pageexec,@zygoloid |

## Extended Description 

A piece of code in the Linux kernel triggers -Warray-bounds when it appears that it should not ([initial report](https://lore.kernel.org/r/YS0nJtNDCwfbaubZ@Ryzen-9-3900X.localdomain/)). Attached is a small reproducer, which has been derived from [arch/powerpc/kernel/signal_32.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/powerpc/kernel/signal_32.c?h=v5.14).

```
$ clang -fsyntax-only -Weverything warray-bounds.c
warray-bounds.c:12:27: warning: array index 3 is past the end of the array (which contains 1 element) [-Warray-bounds]
                __asm__ goto("" : : "r"(oldset->sig[3] >> 32) : : failed);
                                        ^           ~
warray-bounds.c:4:2: note: array 'sig' declared here
        unsigned long sig[NSIG_WORDS];
        ^
warray-bounds.c:16:27: warning: array index 2 is past the end of the array (which contains 1 element) [-Warray-bounds]
                __asm__ goto("" : : "r"(oldset->sig[2] >> 32) : : failed);
                                        ^           ~
warray-bounds.c:4:2: note: array 'sig' declared here
        unsigned long sig[NSIG_WORDS];
        ^
warray-bounds.c:20:27: warning: array index 1 is past the end of the array (which contains 1 element) [-Warray-bounds]
                __asm__ goto("" : : "r"(oldset->sig[1] >> 32) : : failed);
                                        ^           ~
warray-bounds.c:4:2: note: array 'sig' declared here
        unsigned long sig[NSIG_WORDS];
        ^
3 warnings generated.

$ gcc -fsyntax-only -Wall -Wextra -Wpedantic warray-bounds.c
```

This appears similar to https://github.com/llvm/llvm-project/issues/35463 but the second function appears to do the right thing and not warn so I do not think that it is the same situation.

**Comment by nickdesaulniers** [comment]
https://reviews.llvm.org/D116059
